impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java

```diff
@@ -336,48 +345,126 @@ public void clear() throws CacheAccessException {
     return cachingTierRef.get();
   }
 
+  @ServiceDependencies({CompoundCachingTier.Provider.class,
+      OnHeapStore.Provider.class, OffHeapStore.Provider.class, OffHeapDiskStore.Provider.class})
   public static class Provider implements Store.Provider {
 
+    private static final Set<Set<ResourceType.Core>> SUPPORTED_RESOURCE_COMBINATIONS;
+    static {
+      // Logic in setTierConfigurations must mirror this set
+      final Set<Set<ResourceType.Core>> supported = new HashSet<Set<ResourceType.Core>>();
+      supported.add(EnumSet.of(ResourceType.Core.HEAP, ResourceType.Core.DISK));
+      supported.add(EnumSet.of(ResourceType.Core.HEAP, ResourceType.Core.OFFHEAP));
+      supported.add(EnumSet.of(ResourceType.Core.HEAP, ResourceType.Core.OFFHEAP, ResourceType.Core.DISK));
+      SUPPORTED_RESOURCE_COMBINATIONS = supported;
+    }
+
     private volatile ServiceProvider<Service> serviceProvider;
     private final ConcurrentMap<Store<?, ?>, Map.Entry<CachingTier.Provider, AuthoritativeTier.Provider>> providersMap = new ConcurrentWeakIdentityHashMap<Store<?, ?>, Map.Entry<CachingTier.Provider, AuthoritativeTier.Provider>>();
 
     @Override
     public int rank(final Set<ResourceType> resourceTypes, final Collection<ServiceConfiguration<?>> serviceConfigs) {
-      return 0;
+      if (SUPPORTED_RESOURCE_COMBINATIONS.contains(resourceTypes)) {
+        return resourceTypes.size();
+      } else {
+        return 0;
+      }
     }
 
     @Override
     public <K, V> Store<K, V> createStore(Configuration<K, V> storeConfig, ServiceConfiguration<?>... serviceConfigs) {
       CacheStoreServiceConfiguration cacheStoreServiceConfig = findSingletonAmongst(CacheStoreServiceConfiguration.class, (Object[])serviceConfigs);
-      if (cacheStoreServiceConfig == null) {
-        throw new IllegalArgumentException("Cache store cannot be configured without explicit config");
+      if (cacheStoreServiceConfig != null) {
```

Comment:
If I'm reading whats going on here right, would it maybe be simpler to make CacheStoreServiceConfiguration not implement ServiceConfiguration, and then rename it to CacheStoreConfiguration?